### PR TITLE
[6054] Clear dashlet options after performing actions in new Dashboard

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/6054

### CODE REVIEW

**Review by rart — CHANGES_REQUESTED** [review]
This seems possibly disruptive. Events flow through the system at all times and your selection can get lost as you're working. We should try to keep state as steady as possible.

When those events happen, it's ok to fetch but we should keep state steady:
- Stay in the same page, keep selection. You need to fetch up to that page (all prior pages) and check selection is still valid. There are some edge cases, we can refine later.
- Clear the selection after any action is triggered.

Also please default page size to 50 items on all widgets, enable a background refresh, the refresh button should disable (`loading={true}`) whist fetching/reloading.

File: ui/app/src/components/ActivityDashlet/ActivityDashlet.tsx [hunk 1]
```diff
@@ -164,13 +164,16 @@ export function ActivityDashlet(props: ActivityDashletProps) {
...
+    limit: 50,
     offset: 0,
     total: null
   });
   const [selectedActivities, setSelectedActivities, activities] = useSelectionLookupState<ActivitiesAndAll>({
     ALL: true
   });
+  // This is used separately from state.loading because the loading state is used to show loaders (skeleton). This one
+  // still sets to true so elements can be disabled while fetching.
+  const [isFetching, setIsFetching] = useState(false);
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/ActivityDashlet.tsx]: Rename both appropriately. What's the distinction between them though? When is one true vs the other... 
Could/should this be a part of the existing spread state?

File: ui/app/src/components/ActivityDashlet/ActivityDashlet.tsx [hunk 2]
```diff
@@ -285,7 +308,7 @@ export function ActivityDashlet(props: ActivityDashletProps) {
       borderLeftColor={borderLeftColor}
       title={<FormattedMessage id="words.activity" defaultMessage="Activity" />}
       headerAction={
-        <IconButton onClick={() => onRefresh()}>
+        <IconButton onClick={() => onRefresh(true)} disabled={isFetching}>
           <RefreshRounded />
         </IconButton>
```
Comment (rart) [inline ui/app/src/components/ActivityDashlet/ActivityDashlet.tsx]:
```suggestion
        <LoadingButton onClick={() => onRefresh(true)} loading={isFetching} sx={{ borderRadius: '50%', padding: '8px', minWidth: 0 }}>
          <DeleteIcon />
        </LoadingButton>
```

File: ui/app/src/components/ExpiringDashlet/ExpiringDashlet.tsx [hunk 3]
```diff
@@ -140,7 +140,7 @@ export function ExpiringDashlet(props: ExpiringDashletProps) {
       borderLeftColor={borderLeftColor}
       title={<FormattedMessage id="words.expiring" defaultMessage="Expiring" />}
       headerAction={
-        <IconButton onClick={onRefresh}>
+        <IconButton onClick={onRefresh} disabled={state.loading}>
```
Comment (rart) [inline ui/app/src/components/ExpiringDashlet/ExpiringDashlet.tsx]: As above

File: ui/app/src/components/MyRecentActivityDashlet/MyRecentActivityDashlet.tsx [hunk 4]
```diff
@@ -68,29 +68,21 @@ export function MyRecentActivityDashlet(props: MyRecentActivityDashletProps) {
...
-        limit
-      }).subscribe((feed) => {
-        setState({ feed, total: feed.total, offset: 0, loading: false });
-      });
-    },
-    [setState, site, username, limit]
-  );
+  // This is used separately from state.loading because the loading state is used to show loaders (skeleton). This one
+  // still sets to true so elements can be disabled while fetching.
+  const [isFetching, setIsFetching] = useState(false);
```
Comment (rart) [inline ui/app/src/components/MyRecentActivityDashlet/MyRecentActivityDashlet.tsx]: As above

File: ui/app/src/components/MyRecentActivityDashlet/MyRecentActivityDashlet.tsx [hunk 5]
```diff
@@ -167,7 +164,7 @@ export function MyRecentActivityDashlet(props: MyRecentActivityDashletProps) {
         footer: { justifyContent: 'space-between' }
       }}
       headerAction={
-        <IconButton onClick={onRefresh}>
+        <IconButton onClick={onRefresh} disabled={isFetching}>
```
Comment (rart) [inline ui/app/src/components/MyRecentActivityDashlet/MyRecentActivityDashlet.tsx]: As above

**Review by rart — CHANGES_REQUESTED** [review]
Apply changes to all.

File: ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx [hunk 6]
```diff
@@ -162,31 +148,56 @@ export function ScheduledDashlet(props: ScheduledDashletProps) {
...
+      setItemsById({});
+      forkJoin(
+        new Array(pageNumber + 1).fill(null).map((arrayItem, index) => {
+          return fetchScheduled(site, { limit, offset: index * limit });
+        })
+      ).subscribe((scheduledItemsPages) => {
+        const validatedState = getValidatedSelectionState(scheduledItemsPages, selected, limit);
+        setItemsById(validatedState.itemsById);
+        setState(validatedState.state);
+      });
```
Comment (rart) [inline ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx]: Why load pages separately?

File: ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx [hunk 7]
```diff
@@ -75,58 +78,70 @@ export function PendingApprovalDashlet(props: PendingApprovalDashletProps) {
...
+      setItemsById({});
+      forkJoin(
+        new Array(pageNumber + 1).fill(null).map((arrayItem, index) => {
+          return fetchPendingApproval(site, { limit, offset: index * limit });
+        })
+      ).subscribe((pendingApprovalItemsPages) => {
+        const validatedState = getValidatedSelectionState(pendingApprovalItemsPages, selected, limit);
+        setItemsById(validatedState.itemsById);
+        setState(validatedState.state);
+      });
```
Comment (rart) [inline ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx]: Why load pages separately?

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 8]
```diff
@@ -68,57 +70,70 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
...
+      setItemsById({});
+      forkJoin(
+        new Array(pageNumber + 1).fill(null).map((arrayItem, index) => {
+          return fetchUnpublished(site, { limit, offset: index * limit });
+        })
+      ).subscribe((unpublishedItemsPages) => {
+        const validatedState = getValidatedSelectionState(unpublishedItemsPages, selected, limit);
+        setItemsById(validatedState.itemsById);
+        setState(validatedState.state);
+      });
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]: ?

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 9]
```diff
@@ -68,57 +70,70 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
...
+    limit: 50,
     offset: 0
   });
   const currentPage = offset / limit;
   const totalPages = total ? Math.ceil(total / limit) : 0;
-  const [itemsByPath, setItemsByPath] = useSpreadState<LookupTable<DetailedItem>>({});
-  const selectedItems = Object.values(itemsByPath)?.filter((item) => selected[item.id]) ?? [];
+  const [itemsById, setItemsById] = useState<LookupTable<DetailedItem>>({});
+  const itemsByPathRef = useRef({});
+  itemsByPathRef.current = itemsById;
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]:
```suggestion
  const itemsByPathRef = useUpdateRefs(itemsById);
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx [hunk 10]
```diff
@@ -75,58 +77,67 @@ export function PendingApprovalDashlet(props: PendingApprovalDashletProps) {
...
   );
 
+  const loadPagesUntil = useCallback(
+    (pageNumber: number, backgroundRefresh?: boolean) => {
+      setState({
+        loading: true,
+        loadingSkeleton: !backgroundRefresh,
+        ...(!backgroundRefresh && { items: null })
+      });
+      setItemsById({});
```
Comment (rart) [inline ui/app/src/components/PendingApprovalDashlet/PendingApprovalDashlet.tsx]: Necessary?

File: ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx [hunk 11]
```diff
@@ -162,31 +147,53 @@ export function ScheduledDashlet(props: ScheduledDashletProps) {
...
-    onRefresh();
-  }, [onRefresh]);
+  const loadPagesUntil = useCallback(
+    (pageNumber: number, backgroundRefresh?: boolean) => {
+      setState({
+        loading: true,
+        loadingSkeleton: !backgroundRefresh,
+        ...(!loadingSkeleton && { items: null })
+      });
+      setItemsById({});
```
Comment (rart) [inline ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx]: necessary?

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 12]
```diff
@@ -68,57 +70,67 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
...
 
+  const loadPagesUntil = useCallback(
+    (pageNumber: number, backgroundRefresh?: boolean) => {
+      setState({
+        loading: true,
+        loadingSkeleton: !backgroundRefresh,
+        ...(!backgroundRefresh && { items: null })
+      });
+      setItemsById({});
+
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]:
```suggestion
```

File: ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx [hunk 13]
```diff
@@ -154,25 +166,27 @@ export function UnpublishedDashlet(props: UnpublishedDashletProps) {
       items.forEach((item) => {
         itemsObj[item.id] = parseSandBoxItemToDetailedItem(item);
       });
-      setItemsByPath(itemsObj);
+      setItemsById({ ...itemsByPathRef.current, ...itemsObj });
     }
-  }, [items, setItemsByPath]);
+    // itemsByPathRef is a ref, so it's not part of the dependency array
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (rart) [inline ui/app/src/components/UnpublishedDashlet/UnpublishedDashlet.tsx]: Doesn't matter, doesn't hurt to add it and that way we don't loose the linting.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx (hunk 11 above)
Comment (jvega190) [reply ui/app/src/components/ScheduledDashlet/ScheduledDashlet.tsx]: not needed, removing

**Review by rart — APPROVED** [review]

**Comment by sumerjabri** [comment]
@jvega190 please fix conflict

**Review by rart — APPROVED** [review]

**Pull request merged**